sbot/metafeed.go

```diff
@@ -128,59 +192,95 @@ func (s metaFeedsService) CreateSubFeed(purpose string, format refs.RefAlgo) (re
 	if err != nil {
 		return refs.FeedRef{}, err
 	}
-	dbSubkeyID := keys.ID(newSubfeedAsTFK)
 
 	// store the singing key
-	err = s.keys.AddKey(dbSubkeyID, keys.Recipient{
-		Key:    keys.Key(newSubfeedKeyPair.PrivateKey),
-		Scheme: keys.SchemeFeedMessageSigningKey,
-	})
+	err = storeKeyPair(s.keys, newSubfeedKeyPair)
 	if err != nil {
 		return refs.FeedRef{}, err
 	}
 
-	// add the subfeed to the list of keys
-	listData, err := slp.Encode(newSubfeedAsTFK, []byte(purpose))
+	metaPublisher, err := message.OpenPublishLog(s.rxLog, s.users, mountKeyPair, message.SetHMACKey(s.hmacSecret))
 	if err != nil {
 		return refs.FeedRef{}, err
 	}
 
-	subfeedListID := keys.IDFromFeed(s.rootKeyPair.Feed)
-	err = s.keys.AddKey(subfeedListID, keys.Recipient{
-		Key:    keys.Key(listData),
-		Scheme: keys.SchemeMetafeedSubkey,
+	addContent := metamngmt.NewAddDerivedMessage(mountKeyPair.Feed, newSubfeedKeyPair.Feed, purpose, nonce)
 
-		Metadata: keys.Metadata{
-			ForFeed: newSubfeedKeyPair.Feed,
-		},
-	})
+	// we had some metadata :^)
+	if len(optionalMetadata) > 0 {
+		if len(optionalMetadata) > 1 {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: metadata to contain a single map, not multiple")
+		}
+		metadata := optionalMetadata[0]
+		// note (2021-09-13): we're current only supporting ssb-ql-0 as a querylang
+		if _, exists := metadata["querylang"]; !exists {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata to have key `querylang`")
+		} else if metadata["querylang"] != "ssb-ql-0" {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata key `querylang` to have value `ssb-ql-0`, not `%s`", metadata["querylang"])
+		}
+
+		if _, exists := metadata["type"]; !exists {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata to have key `type`")
+		}
+		if _, exists := metadata["author"]; !exists {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata to have key `author`")
+		}
+
+		// create a `query` object with the expected ssb-ql-0 format
+		queryJson, err := json.Marshal(metadataQuery{Author: metadata["author"], Type: metadata["type"]})
```

Comment:
add `private: false`

Reply:
haha yeah, that's what i was thinking as well & why i left it out :)